tests/unit/test_task_storage.py

```diff
@@ -0,0 +1,160 @@
+
+
+class SimpleInMemoryTask(TaskProtocol):
+    def __init__(self):
+        self.task_id = str(uuid.uuid1())
+        self.endpoint = None
+        self.status = TaskState.RECEIVED
+        self.result = None
+        self.result_reference = None
+
+
+def test_memory_storage_simple():
+    memstore = MemoryTaskStorage()
+    result = "foo result"
+    task = SimpleInMemoryTask()
+
+    b = memstore.store_result(task, result)
+    assert b is not None
+    assert memstore.get_result(task) == result
+
+
+def test_chained_storage_first_success():
+    memstore1 = MemoryTaskStorage()
+    memstore2 = MemoryTaskStorage()
+
+    chain = ChainedTaskStorage(memstore1, memstore2)
+
+    result = "foo result"
+    task = SimpleInMemoryTask()
+
+    b = chain.store_result(task, result)
+    assert b is True
+    assert memstore2.get_result(task) is None
+    assert memstore1.get_result(task) == result
+    assert chain.get_result(task) == result
+
+
+# @pytest.mark.skip
+def test_thresholded_storage_limit():
+    store = ThresholdedMemoryTaskStorage(result_limit_chars=10)
+
+    result = "result"
+    task = SimpleInMemoryTask()
+
+    b = store.store_result(task, result)
+    assert b is True
+    assert store.get_result(task) == result
+
+    result2 = "result too long for char limit"
+    task2 = SimpleInMemoryTask()
+
+    with pytest.raises(StorageException):
+        b = store.store_result(task2, result2)
+    with pytest.raises(StorageException):
+        store.get_result(task2)
+
+
+def test_chained_with_threshold():
+    memstore1 = ThresholdedMemoryTaskStorage(result_limit_chars=3)
+    memstore2 = MemoryTaskStorage()
+
+    chain = ChainedTaskStorage(memstore1, memstore2)
+
+    result = "foo result"
+    task = SimpleInMemoryTask()
+
+    b = chain.store_result(task, result)
+    assert b is True
+    with pytest.raises(StorageException):
+        memstore1.get_result(task)
+    assert memstore2.get_result(task) == result
+    assert chain.get_result(task) == result
+
+
+def test_null_storage():
+    store = NullTaskStorage()
+
+    result = "result"
+    task = SimpleInMemoryTask()
+
+    with pytest.raises(StorageException):
+        store.store_result(task, result)
+
+    with pytest.raises(StorageException):
+        store.get_result(task)
+
+
+def test_failing_chain_storage():
+    store1 = NullTaskStorage()
+    store2 = NullTaskStorage()
+
+    chain = ChainedTaskStorage(store1, store2)
+
+    result = "result"
+    task = SimpleInMemoryTask()
+
+    with pytest.raises(StorageException):
+        chain.store_result(task, result)
+    # maybe_todo : Figure out what the behavior here should be
+    # if the storage fails, should it be possible to retrieve
+    # the result?
+    # with pytest.raises(StorageException):
+    #    chain.get_result(task)
+
+
+def test_s3_task_storage():
+    store = S3TaskStorage("funcx-test-1")
+
+    task = SimpleInMemoryTask()
+    result = "Hello World!"
+    store.store_result(task, result)
```

Comment:
Could we mock boto3 and actually test the operation is using the bucket and object name we expect?

Reply:
Ben, I'm thinking of using this -> https://botocore.amazonaws.com/v1/documentation/api/latest/reference/stubber.html. For now the tests depend on actually talking to S3 so I've marked them to skip if AWS_PROFILE is not set. The plan is to replicate some of these tests to the functional category and have the s3 client mocked for unit tests. On the functional side, I'll add tests to confirm that our system is in fact interacting with S3 correctly.